primitives/utility/src/lib.rs

```diff
@@ -51,3 +61,191 @@ impl<T: UpwardMessageSender, W: WrapVersion> SendXcm for ParentAsUmp<T, W> {
 		}
 	}
 }
+
+/// Contains information to handle refund/payment for xcm-execution
+#[derive(Clone, Eq, PartialEq, Debug)]
+struct AssetTraderRefunder<Balance, AssetId> {
+	// The amount of weight bought minus the weigh already refunded
+	weight_outstanding: Weight,
+	// The balance paid for the weight bought minus the balance already refunded
+	outstanding_balance: Balance,
+	// The multilocation representing the fungible asset in which fees were paid
+	location: MultiLocation,
+	// The local assetId the aforementioned location corresponds to
+	asset: AssetId,
+}
+
+/// Charges for exercution in the first multiasset of those selected for fee payment
+/// Only succeeds for Concrete Fungible Assets
```

Comment:
So the limitation here is purely because you store an `Option` field in your struct, correct? If so, perhaps we can just store a `Vec` instead? Because I can see that it's totally valid to use multiple `BuyExecution`s when one asset isn't enough to pay for fees, and another asset is provided to fill in the remaining deficit.

Reply:
Right now the only thing that I know of making sure you are buying enough weight is the `AllowTopLevelPaidExecutionFrom` barrier. Hence inserting a second  `BuyExecution` would not help for anything but giving away tokens (but I might be missing something)

Reply:
Although again this is making an assumption on a specific barrier being implemented. However, I find it hard to handle the `refund` case when you have multiple assets, in which one of them should we refund?

Reply:
Hmm, so having this extra functionality does mean that `AllowTopLevelPaidExecutionsFrom` would have to change a bit in order to check whether other assets is enough to pay for the weights. I do not know if looking up other assets entail a database lookup, but if it does, then it's an additional problem because barrier conditions need to be lightweight to check, otherwise it's an attack vector.

In either case, this looks like a really tough problem to solve, so I don't mind if we simply restrict it to 1 `BuyExecution` for now and move on from here.

Reply:
perfect, I agree this is a tough problem with respect to DB reads